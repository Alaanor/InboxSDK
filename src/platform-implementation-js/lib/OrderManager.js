/* @flow */

import sortedIndexBy from 'lodash/sortedIndexBy';

export type Item<T> = {
  groupId: string;
  id: string;
  orderHint: number;
  value: T;
};
export default class OrderManager<T> {
  _items: Array<Item<T>> = [];
  _storage: Storage;
  constructor(storage: Storage=window.localStorage) {
    this._storage = storage;
  }
  addItem(item: Item<T>) {
    const itemsInSameGroup = this._items.filter(x => x.groupId === item.groupId);
    if (itemsInSameGroup.length === 0) {
      if (Math.random() < 0.5) {
        this._items = [item].concat(this._items);
      } else {
        this._items = this._items.concat([item]);
      }
    } else {
      const ixInGroup = sortedIndexBy(itemsInSameGroup, item, x => x.orderHint);
      let ix;
      if (ixInGroup === 0) {
        const afterItem = itemsInSameGroup[0];
        ix = this._items.indexOf(afterItem);
      } else {
        const beforeItem = itemsInSameGroup[ixInGroup-1];
        ix = this._items.indexOf(beforeItem)+1;
      }
      this._items = this._items.slice(0, ix).concat([item], this._items.slice(ix));
    }
  }
  removeItem(id: string) {
    this._items = this._items.filter(item => item.id !== id);
  }
  getOrderedItems(): Array<Item<T>> {
    return this._items;
  }
}
