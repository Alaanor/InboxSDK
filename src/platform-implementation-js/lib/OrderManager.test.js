/* @flow */

import OrderManager from './OrderManager';
import seed from 'seed-random';
import times from 'lodash/times';
import MockStorage from 'mock-webstorage';

afterEach(() => {
  seed.resetGlobal();
});

test('addItem, removeItem', () => {
  const storage: Object = new MockStorage();
  const o = new OrderManager(storage);

  const first = o.getOrderedItems();
  expect(first).toEqual([]);
  o.addItem({
    groupId: 'blah',
    id: 'foo',
    orderHint: 0,
    value: {v: 'foo'}
  });

  const second = o.getOrderedItems();
  expect(second).toEqual([{
    groupId: 'blah',
    id: 'foo',
    orderHint: 0,
    value: {v: 'foo'}
  }]);

  // Make sure we didn't mutate the returned array.
  expect(first).toEqual([]);

  o.removeItem('bar');
  expect(o.getOrderedItems().length).toBe(1);
  o.removeItem('foo');
  expect(o.getOrderedItems().length).toBe(0);

  // Make sure we didn't mutate the returned array.
  expect(second.length).toBe(1);
});

test('orderHint within group is respected', () => {
  const storage: Object = new MockStorage();
  const o = new OrderManager(storage);
  o.addItem({
    groupId: 'planets',
    id: 'mercury',
    orderHint: -1,
    value: {v: 'Mercury'}
  });
  o.addItem({
    groupId: 'planets',
    id: 'earth',
    orderHint: 1,
    value: {v: 'Earth'}
  });
  o.addItem({
    groupId: 'planets',
    id: 'venus',
    orderHint: 0,
    value: {v: 'Venus'}
  });
  o.addItem({
    groupId: 'planets',
    id: 'mars',
    orderHint: 1.1,
    value: {v: 'Mars'}
  });
  expect(o.getOrderedItems().map(x => x.id))
    .toEqual(['mercury', 'venus', 'earth', 'mars']);
});

test('orderHint is not respected across groups', () => {
  const orderedIdsOverMultipleRuns = times(10).map(i => {
    seed(`seed ${i}`, {global: true});

    const storage: Object = new MockStorage();
    const o = new OrderManager(storage);
    o.addItem({
      groupId: 'planets',
      id: 'mercury',
      orderHint: -1,
      value: {v: 'Mercury'}
    });
    o.addItem({
      groupId: 'letters',
      id: 'a',
      orderHint: 0,
      value: {v: 'A'}
    });
    return o.getOrderedItems().map(x => x.id);
  });
  expect(orderedIdsOverMultipleRuns).toMatchSnapshot();
});

test('orderHint is respected within groups but not across groups', () => {
  const orderedIdsOverMultipleRuns = times(10).map(i => {
    seed(`seed ${i}`, {global: true});

    const storage: Object = new MockStorage();
    const o = new OrderManager(storage);
    o.addItem({
      groupId: 'letters',
      id: 'b',
      orderHint: 1,
      value: {v: 'B'}
    });
    o.addItem({
      groupId: 'planets',
      id: 'mercury',
      orderHint: -1,
      value: {v: 'Mercury'}
    });
    o.addItem({
      groupId: 'planets',
      id: 'venus',
      orderHint: 0,
      value: {v: 'Venus'}
    });
    o.addItem({
      groupId: 'letters',
      id: 'a',
      orderHint: 0,
      value: {v: 'A'}
    });

    const orderedItems = o.getOrderedItems();
    expect(orderedItems.findIndex(x => x.id === 'mercury'))
      .toBe(orderedItems.findIndex(x => x.id === 'venus') - 1);
    expect(orderedItems.findIndex(x => x.id === 'a'))
      .toBe(orderedItems.findIndex(x => x.id === 'b') - 1);

    return orderedItems.map(x => x.id);
  });
  expect(orderedIdsOverMultipleRuns).toMatchSnapshot();
});
