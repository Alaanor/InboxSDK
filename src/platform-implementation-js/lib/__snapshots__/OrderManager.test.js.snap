exports[`test orderHint is not respected across groups 1`] = `
Array [
  Array [
    "mercury",
    "a",
  ],
  Array [
    "mercury",
    "a",
  ],
  Array [
    "a",
    "mercury",
  ],
  Array [
    "mercury",
    "a",
  ],
  Array [
    "a",
    "mercury",
  ],
  Array [
    "a",
    "mercury",
  ],
  Array [
    "a",
    "mercury",
  ],
  Array [
    "a",
    "mercury",
  ],
  Array [
    "mercury",
    "a",
  ],
  Array [
    "mercury",
    "a",
  ],
]
`;

exports[`test orderHint is respected within groups but not across groups 1`] = `
Array [
  Array [
    "a",
    "b",
    "mercury",
    "venus",
  ],
  Array [
    "a",
    "b",
    "mercury",
    "venus",
  ],
  Array [
    "mercury",
    "venus",
    "a",
    "b",
  ],
  Array [
    "a",
    "b",
    "mercury",
    "venus",
  ],
  Array [
    "mercury",
    "venus",
    "a",
    "b",
  ],
  Array [
    "mercury",
    "venus",
    "a",
    "b",
  ],
  Array [
    "mercury",
    "venus",
    "a",
    "b",
  ],
  Array [
    "mercury",
    "venus",
    "a",
    "b",
  ],
  Array [
    "a",
    "b",
    "mercury",
    "venus",
  ],
  Array [
    "a",
    "b",
    "mercury",
    "venus",
  ],
]
`;
